scheduler/preemption.go

```diff
@@ -0,0 +1,495 @@
+
+	// Subtract the reserved resources of the node
+	if node.ComparableReservedResources() != nil {
+		nodeRemainingResources.Subtract(node.ComparableReservedResources())
+	}
+
+	// Subtract current allocations
+	for _, alloc := range current {
+		nodeRemainingResources.Subtract(alloc.ComparableResources())
+	}
+
+	// Iterate over allocations grouped by priority to find preemptible allocations
+	for _, allocGrp := range allocsByPriority {
+		for len(allocGrp.allocs) > 0 && !allRequirementsMet {
+			closestAllocIndex := -1
+			bestDistance := math.MaxFloat64
+			// find the alloc with the closest distance
+			for index, alloc := range allocGrp.allocs {
+				currentPreemptionCount := computeCurrentPreemptions(alloc, currentPreemptions)
+				maxParallel := 0
+				tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+				if tg != nil && tg.Migrate != nil {
+					maxParallel = tg.Migrate.MaxParallel
+				}
+				distance := getPreemptionScoreForTaskGroupResources(resourcesNeeded, alloc.ComparableResources(), maxParallel, currentPreemptionCount)
+				if distance < bestDistance {
+					bestDistance = distance
+					closestAllocIndex = index
+				}
+			}
+			closestAlloc := allocGrp.allocs[closestAllocIndex]
+
+			if preemptedResources == nil {
+				preemptedResources = closestAlloc.ComparableResources()
+			} else {
+				preemptedResources.Add(closestAlloc.ComparableResources())
+			}
+			availableResources := preemptedResources.Copy()
+			availableResources.Add(nodeRemainingResources)
+
+			// This step needs the original resources asked for as the second arg, can't use the running total
+			allRequirementsMet, _ = availableResources.Superset(resourceAsk.Comparable())
+
+			bestAllocs = append(bestAllocs, closestAlloc)
+
+			allocGrp.allocs[closestAllocIndex] = allocGrp.allocs[len(allocGrp.allocs)-1]
+			allocGrp.allocs = allocGrp.allocs[:len(allocGrp.allocs)-1]
+
+			// this is the remaining total of resources needed
+			resourcesNeeded.Subtract(closestAlloc.ComparableResources())
+		}
+		if allRequirementsMet {
+			break
+		}
+	}
+
+	// Early return if all allocs examined and requirements were not met
+	if !allRequirementsMet {
+		return nil
+	}
+
+	resourcesNeeded = resourceAsk.Comparable()
+	// We do another pass to eliminate unnecessary preemptions
+	// This filters out allocs whose resources are already covered by another alloc
+
+	// Sort bestAllocs by distance descending (without penalty)
+	sort.Slice(bestAllocs, func(i, j int) bool {
+		distance1 := basicResourceDistance(resourcesNeeded, bestAllocs[i].ComparableResources())
+		distance2 := basicResourceDistance(resourcesNeeded, bestAllocs[j].ComparableResources())
+		return distance1 > distance2
+	})
+
+	filteredBestAllocs := eliminateSuperSetAllocationsForTaskGroup(bestAllocs, nodeRemainingResources, resourcesNeeded)
+	return filteredBestAllocs
+
+}
+
+// computeCurrentPreemptions counts the number of other allocations being preempted that match the job and task group of
```

Comment:
todo this could be calculated ahead of time per jobid in the set of existing allocations on the node